examples/scene/isocurve_updates.py

```diff
@@ -106,7 +106,7 @@ def update(ev):
     if index > 0 and index < 25:
         # update left panes rolling upwards
         noise = np.roll(noise, 1, axis=0)
-        image1.set_data(noise)
+        image1.set_data(noise, copy=True)
```

Comment:
This was needed to fix a test failure; the reason why it was working before, is because we were doing unnecessary copies which resulted in `copy=False` not mattering here. But now that we respect the `copy` argument better, this example was failing because `noise` was getting changed between this line and then next.

Reply:
I'm not sure I understand. So one of the Visuals (the image?) was modifying the `noise` array? Doesn't that suggest that it wasn't copying internally when it should have because the copying was necessary to not modify the input data, right?

Reply:
Kinda the opposite: note that by default `copy=False`, and despite that `noise` was clearly getting copied or the example would have crashed. With the current changes, the `copy` kwarg is getting respected, so without `copy=True` we get the failure.

Reply:
So wait, with `copy=False` (the default), who is modifying `noise` in this example? Shouldn't the Visual copy if it is going to modify the data? Or does `copy=False` in these Visuals that it will modify the data inplace no matter what?

Reply:
> So wait, with `copy=False` (the default), who is modifying `noise` in this example?

`image1`

> Shouldn't the Visual copy if it is going to modify the data? Or does `copy=False` in these Visuals that it will modify the data inplace no matter what?

It will modify in place no matter what if it can.

I think the confusion here comes from 2 different use cases for specifying `copy=True`:
1. "Please copy it because I might change it externally before you send it to the gpu and I want you to send THIS EXACT data" (this is the use specified in the docs)
2. "Please copy it before altering it because I will need it elsewhere"

This is case 2.

Reply:
And in what parts of the code is it modifying it in place? I don't think it ever used to copy unless it needed to. For CPU scaling it always made an implicit copy by doing `data = data * 1. / scale_factor`. I don't know if the tests ever explicitly checked that. Having it modify things in place seems wrong, but if that's how it has always been then I guess we leave it.

@almarklein @tlambert03 @kmuehlbauer Any opinion on this? Should `ImageVisual.set_data(data, copy=False)` be allowed to modify the data inplace?

Reply:
> but if that's how it has always been then I guess we leave it.

Well, it really hasn't, that's why the test used to pass :P The problem is that there was no *explicit* check that said "are we going to modify? If yes, let's copy". I can see the benefits though, so I'll let others chip in with more opinions.

Reply:
I don't think the copy inside the Visual/Texture needs to be explicit, but the check to make sure it isn't modified should be tested.

Reply:
Mhm, I think I tracked down the exact issue. In `Image.set_data` only some basic checks happen, and then the data is saved as is in `Image._data` for deferred uploading to the GPU. So, no matter how careful the texture itself is at not changing the data, if the data gets changed on the outside *before* `Image._prepare_draw` happens, the data received by the texture will be already messed up.

Not sure if there's a way around it other than immediately processing the data on the texture side, or simply passing `copy=True`.

Reply:
But what you're describing is case 1 that you listed above. You said this fix was for case 2. I'd have to find the old discussions but I don't think we ever want the Visual to modify the users provided array. The user can tell us to copy the data to avoid corruption (case 1), but otherwise the data should be copied whenever necessary.

Reply:
Man this is finnicky to debug... It seems that the thing that's actually failing the line visual below instead...

Reply:
Upon more testing, It turns out that this bug is unrelated to this PR. The same exact behaviour happens on main; however, it's very inconsistent to reproduce (both on main and here). To reproduce, simply run the `isocurve_updates` example a few times, and you'll see the failure.

Reply:
What kind of failure? I've run it a couple times now and I'm not sure what to expect.

Reply:
[This is the failure on CI](https://github.com/vispy/vispy/runs/7538804305?check_suite_focus=true#step:11:178), and this is how it looks like on my machine:

https://user-images.githubusercontent.com/23482191/182164048-ba1c4275-3bb0-43b0-99e9-5bbca7ef295f.mp4

Reply:
Those look like very different errors between CI and your machine. On `main`, you get this without any changes?

Reply:
The example also seems to create `img_data1` but then never uses it.

Reply:
I also noticed that in your screencast that the bottom two boxes are the ones that don't display at all, but those are only Isocurves, not the ImageVisuals created in the example.

Reply:
Lol, you're right, thanks for double checkin... I could swear I saw the same error. I must be wrong though, I can't reproduce anymore. Maybe I was on this branch and though it was main :facepalm:

The error above is indeed happening on main without any changes, but happens regularly if I rerun the example in the same ipython session. Definitely a bug, and definitely unrelated.

Reply:
Ok, looking at the diff now I see how we raised an error before this PR if updating couldn't be done without copying. I think copying when we have to is a reasonable assumption, but we may have to triple check the docstrings for the ImageVisual and the textures. Looks good to me though.

Reply:
...except for the test failures :wink: